### ISSUE TITLE
Missing `Access-Control-Allow-Origin` header from responses.

### ISSUE THREAD

**Issue by Seldszar** [speedruncomorg/api#52]
I would like to consume the API but it's currently impossible from any browser.
This issue is due to the missing `Access-Control-Allow-Origin` header.

Would you please add this header?
Thanks in advance.

**Comment by TsundereBug** [comment]
My browser has no issues with browsing through the API. Would help to know what Operating System and Browser you are using.

**Comment by ROMaster2** [comment]
I've had this problem too. Opening up the API directly works just fine, but using Javascript to access the API runs into that.

**Comment by TsundereBug** [comment]
If it's JavaScript from a webpage, that's probably due to Cross Origin requests? I have had trouble like that in the past.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS
Access-Control-Allow-Origin - MDN - Mozilla

**Comment by xTVaser** [comment]
While it would be ideal for the header to be added, there is a workaround that is mentioned in the documentation using JSONP - https://github.com/speedruncomorg/api/tree/master/version1

For example:
```javascript
function printResults(json) {
    console.log(json)
}

$("#searchBtn").click(function(){
    $.ajax({
        url: "https://www.speedrun.com/api/v1/users?lookup=vaser",
        dataType: "jsonp",
        jsonpCallback: "printResults"
    });
})
```

You can also use the ?callback=<function name> parameter if you'd like

**Comment by xTVaser** [comment]
Just ran into a rather large problem due to not having the request headers.  I've been working on making a Twitch Extension using the API and since one of the restrictions is that all CSS and JS has to be loaded from Twitch's CDN, JSONP is not allowed.

This means that doing any client-side work with the API is impossible in this use-case unless the header is added....or all of the work can be shifted server-side which has obvious performance/scaling issues.

**Comment by TsundereBug** [comment]
Try using Google Ajax for now? I heard it can bypass CORS, not so sure.

This issue really needs to be fixed.

**Comment by xTVaser** [comment]
Hmm, can't say I've ever heard of Google AJAX, tried searching around but I don't think I'm finding the right thing, If you have a link I'll definitely take a look.  Thanks

**Comment by TsundereBug** [comment]
https://stackoverflow.com/questions/3076414/ways-to-circumvent-the-same-origin-policy
javascript - Ways to circumvent the same-origin policy - Stack Overflow
https://stackoverflow.com/questions/29837162/bypassing-cors-with-php-and-javascript-ajax
Bypassing CORS with PHP and Javascript/Ajax - Stack Overflow
https://gist.github.com/karanth/8467944
Notes on JSONP and CORS - cross-domain AJAX calls · GitHub
https://www.thepolyglotdeveloper.com/2014/08/bypass-cors-errors-testing-apis-locally/
Bypass CORS Errors When Testing APIs Locally

**Comment by xTVaser** [comment]
Thanks for those links, I've read through them and ya JSONP does circumvent CORS issues and I was using it successfully.  But the issue with Twitch is a CSP violation since JSONP is basically injecting arbitrary javascript.  

However if CORS was enabled for the API, there would be no need to use JSONP and inject javascript, it would just be a normal XHR.  So I think the best thing to do in my case is to just wait and hope this gets fixed.

**Comment by sgt-kabukiman** [comment]
Thanks for all your comments. :)

Given that we require an API key to authenticate, I see no CSRF issue in allowing cross site requests to the API.

Please check if the API now works correctly in your specific usecases, as I've just added the neccessary CORS headers to allow requests from everywhere. Given the somewhat "old" API codebase, there might be some bugs though ;-)

**Comment by xTVaser** [comment]
Thanks for taking a look at this!

I just switched everything over from using JSONP to normal requests and ran into zero issues, so everything looks good from my end so far. +💯

**Comment by xTVaser** [comment]
Just an update on this, while testing I have noticed every now and again that the API is throwing CORS issues back.  A refresh/retry usually will usually remedy this, so maybe this is caused by rate-limiting?

**Comment by ROMaster2** [comment]
There's a rate limit of 100 requests per second.

**Comment by s5bug** [comment]
CORS should still be sent on 429s.

**Comment by xTVaser** [comment]
Yes I'm aware of the limit, it was just the first thing that came to my mind since it appears to be an intermittent problem.  I no longer think this because the problem reappeared after several minutes of down-time upon first attempt.

Also, is the rate limit actually 100 requests per *second*? Documentation states 100 requests per minute https://github.com/speedruncomorg/api/blob/master/throttling.md and that would explain why I've never actually ran into rate-limiting issues when I feel like I'm doing way more than 100 per minute.

**Comment by ROMaster2** [comment]
minute

**Comment by xTVaser** [comment]
Just so an example of the CORS issue is documented here.  Same intermittent problem described above that is typically always resolved by a retry. 
![image](https://user-images.githubusercontent.com/13153231/34912173-0eb4846a-f8a8-11e7-8a8b-2b96bd578cd2.png)

**Comment by xTVaser** [comment]
Did some testing and it appears that the API is frequently returning `502`s which **do not** have the CORS headers (hence the CORS violations), so there is two problems at play.

Heres a JS Fiddle, if you run it and look in the network tab in the dev tools you can see the failed requests  that happen quite often.  https://jsfiddle.net/ro5pdpak/11/

![image](https://user-images.githubusercontent.com/13153231/34921174-8a9df738-f94c-11e7-8c3a-1a0632d3bfcf.png)

I should probably make my app more resilent and re-try requests but that would be much easier to do with a proper response instead of just getting knocked down due to CORS.  Perhaps this is due to the usage spike from GDQ, however I saw this problem weeks before the event so maybe the API just needs some more resources as I also see that the latency has gone up quite significantly (was almost always sub-second before, now it is often multiple-seconds).
